flink-clients/src/test/java/org/apache/flink/client/deployment/ClusterClientServiceLoaderTest.java

```diff
@@ -0,0 +1,167 @@
+		serviceLoaderUnderTest = new DefaultClusterClientServiceLoader();
+	}
+
+	@Test
+	public void testStandaloneClusterClientFactoryDiscovery() {
+		final Configuration config = new Configuration();
+		config.setString(DeploymentOptions.TARGET, StandaloneClientFactory.ID);
+
+		ClusterClientFactory<StandaloneClusterId> factory = serviceLoaderUnderTest.getClusterClientFactory(config);
+		assertTrue(factory instanceof StandaloneClientFactory);
+	}
+
+	@Test
+	public void testFactoryDiscovery() {
+		final Configuration config = new Configuration();
+		config.setString(DeploymentOptions.TARGET, VALID_TARGET);
+
+		final ClusterClientFactory<Integer> factory = serviceLoaderUnderTest.getClusterClientFactory(config);
+		assertNotNull(factory);
+
+		final Integer id = factory.getClusterId(config);
+		assertThat(id, allOf(is(notNullValue()), equalTo(VALID_ID)));
+	}
+
+	@Test(expected = IllegalStateException.class)
+	public void testMoreThanOneCompatibleFactoriesException() {
+		final Configuration config = new Configuration();
+		config.setString(DeploymentOptions.TARGET, AMBIGUOUS_TARGET);
+
+		serviceLoaderUnderTest.getClusterClientFactory(config);
+		fail();
+	}
+
+	@Test
+	public void testNoFactoriesFound() {
+		final Configuration config = new Configuration();
+		config.setString(DeploymentOptions.TARGET, NON_EXISTING_TARGET);
+
+		final ClusterClientFactory<Integer> factory = serviceLoaderUnderTest.getClusterClientFactory(config);
+		assertNull(factory);
+	}
+
+	/**
+	 * Test {@link ClusterClientFactory} that is successfully discovered.
+	 */
+	public static class ValidClusterClientFactory extends DummyClusterClientFactory {
+
+		public static final String ID = VALID_TARGET;
+
+		@Override
+		public boolean isCompatibleWith(Configuration configuration) {
+			return configuration.getString(DeploymentOptions.TARGET).equals(VALID_TARGET);
+		}
+
+		@Nullable
+		@Override
+		public Integer getClusterId(Configuration configuration) {
+			return VALID_ID;
+		}
+	}
+
+	/**
+	 * Test {@link ClusterClientFactory} that has a duplicate.
+	 */
+	public static class FirstCollidingClusterClientFactory extends DummyClusterClientFactory {
+
+		public static final String ID = AMBIGUOUS_TARGET;
+
+		@Override
+		public boolean isCompatibleWith(Configuration configuration) {
+			return configuration.getString(DeploymentOptions.TARGET).equals(AMBIGUOUS_TARGET);
+		}
+	}
+
+	/**
+	 * Test {@link ClusterClientFactory} that has a duplicate.
+	 */
+	public static class SecondCollidingClusterClientFactory extends DummyClusterClientFactory {
+
+		public static final String ID = AMBIGUOUS_TARGET;
+
+		@Override
+		public boolean isCompatibleWith(Configuration configuration) {
+			return configuration.getString(DeploymentOptions.TARGET).equals(AMBIGUOUS_TARGET);
+		}
+	}
+
+	/**
+	 * A base test {@link ClusterClientFactory} that supports no operation and is meant to be extended.
+	 */
+	public static class DummyClusterClientFactory implements ClusterClientFactory<Integer> {
```

Comment:
I am afraid we already have another `DummyClusterClientFactory`. This one could be a settable cluster client factory so that we inject lambda as function to be applied and if the function is `null`, throws `UnsupportedOperationException`. It will simplify inheritance also.

Reply:
Hmmm I think you make use of `ServerLoader` so we  have to use different classes. The naming issue is still valid but let me think of this loading strategy.

Reply:
Hi @TisonKun , I am not sure what you mean here. You mean that we can merge the two classes to avoid code duplication?

Reply:
I think these two `DummyClusterClientFactory` are different implementations so that they are unable to be merged. This once can be called `TestingClusterClientFactory` or something for avoiding name conflict. Besides, it is a abstract base class for testing dynamic loading IMO.

Reply:
I see. I will rename.